File: rheosim-backend/rheosim-infrastructure/src/main/java/com/rheosim/infrastructure/simulation/engine/KelvinVoigtLaw.java
package com.rheosim.infrastructure.simulation.engine;

import com.rheosim.domain.project.model.ConstitutiveModelType;
import com.rheosim.domain.simulation.port.ConstitutiveLaw;
import org.springframework.stereotype.Component;

@Component
public class KelvinVoigtLaw implements ConstitutiveLaw {

    @Override
    public ConstitutiveModelType getModelType() {
        return ConstitutiveModelType.KELVIN_VOIGT;
    }

    @Override
    public double computeRelaxationModulus(double time, double[] parameters) {
        // Kelvin-Voigt: G(t) = G + eta * delta(t) — not well-defined for continuous relaxation
        // Approximation: G(t) = G (constant, since the element doesn't relax)
        double g = parameters[0];
        return g;
    }

    @Override
    public double computeCreepCompliance(double time, double[] parameters) {
        // J(t) = (1/G) * (1 - exp(-t / tau)) where tau = eta/G
        double g = parameters[0];
        double eta = parameters[1];
        double tau = eta / g;
        return (1.0 / g) * (1.0 - Math.exp(-time / tau));
    }

    @Override
    public double computeStorageModulus(double omega, double[] parameters) {
        // G'(omega) = G
        return parameters[0];
    }

    @Override
    public double computeLossModulus(double omega, double[] parameters) {
        // G''(omega) = eta * omega
        double eta = parameters[1];
        return eta * omega;
    }

    @Override
    public double computeComplexViscosity(double omega, double[] parameters) {
        double gPrime = computeStorageModulus(omega, parameters);
        double gDoublePrime = computeLossModulus(omega, parameters);
        return Math.sqrt(gPrime * gPrime + gDoublePrime * gDoublePrime) / omega;
    }

    @Override
    public String[] getParameterNames() {
        return new String[]{"G (Pa)", "eta (Pa.s)"};
    }

    @Override
    public double[] getDefaultParameters() {
        return new double[]{1000.0, 100.0};
    }

    @Override
    public double[] getLowerBounds() {
        return new double[]{1e-3, 1e-6};
    }

    @Override
    public double[] getUpperBounds() {
        return new double[]{1e12, 1e12};
    }

    @Override
    public int getParameterCount(int numberOfBranches) {
        return 2;
    }
}

File: rheosim-compute/src/kelvin_voigt_law.cpp
#include "rheosim/kelvin_voigt_law.h"
#include <cmath>

namespace rheosim {

double KelvinVoigtLaw::compute_relaxation_modulus(double /*time*/, const Eigen::VectorXd& params) const {
    // KV element doesn't relax in pure form; returns spring constant
    return params(0);
}

double KelvinVoigtLaw::compute_creep_compliance(double time, const Eigen::VectorXd& params) const {
    double G = params(0);
    double tau = params(1);
    return (1.0 / G) * (1.0 - std::exp(-time / tau));
}

double KelvinVoigtLaw::compute_storage_modulus(double /*omega*/, const Eigen::VectorXd& params) const {
    return params(0);
}

double KelvinVoigtLaw::compute_loss_modulus(double omega, const Eigen::VectorXd& params) const {
    double G = params(0);
    double tau = params(1);
    return G * omega * tau;
}

double KelvinVoigtLaw::compute_complex_viscosity(double omega, const Eigen::VectorXd& params) const {
    double gp = compute_storage_modulus(omega, params);
    double gpp = compute_loss_modulus(omega, params);
    return std::sqrt(gp * gp + gpp * gpp) / omega;
}

Eigen::VectorXd KelvinVoigtLaw::default_lower_bounds() const {
    Eigen::VectorXd lb(2);
    lb << 1e-3, 1e-10;
    return lb;
}

Eigen::VectorXd KelvinVoigtLaw::default_upper_bounds() const {
    Eigen::VectorXd ub(2);
    ub << 1e12, 1e6;
    return ub;
}

} // namespace rheosim
